### PR TITLE
Allow windows, windows_core (and several other crates) to be compiled and used in no_std environments.

### PR DESCRIPTION
**Pull request by sivadeilra**
This allows many of the crates in this repo to be compiled and used in `#![no_std]` environments. This is a requirement for some low-level components, especially within the implementation of Windows itself.

These crates are updated: `windows`, `windows-core`, `windows-registry`, `windows-result`.  All of these crates now define a new feature, `std`, which is enabled by default.  If you _disable_ the `std` feature, then the crate will be compiled with `#![no_std]`. Because the `std` features are enabled by default, existing users of these crates should not see any difference in behavior.

Nearly all of the changes are extremely simple and do not change actual codegen. These just convert `std::ptr::null_mut` with `core::ptr::null_mut`, etc.

For those cases where a crate used dynamic memory containers (Box, String, Vec), I added an explicit dependency on the `alloc` crate and imported the needed types.  These are the same types that are re-exported from `std`, so this also has no effect on codegen.

In a few cases, code was using things from `std` that are currently only available in `std`, such as `OsString` or `Mutex`. In those cases, I placed the code using those items under `#[cfg(feature = "std")]`.  Since the `std` feature is enabled by default, this will not break any existing users of those crates.

I also noticed a bug in passing in `BasicString`. It can construct a `&[u16]` over a null pointer, which is UB.  I've fixed that bug in this PR.

### CODE REVIEW

File: crates/libs/result/src/strings.rs [hunk 1]
```diff
@@ -35,7 +35,12 @@ impl BasicString {
     }
 
     pub fn as_wide(&self) -> &[u16] {
-        unsafe { std::slice::from_raw_parts(self.as_ptr(), self.len()) }
```
Comment (sivadeilra) [inline crates/libs/result/src/strings.rs:38]: This is UB, in safe code, if the string is a null pointer.

File: crates/libs/registry/Cargo.toml [hunk 2]
```diff
@@ -11,6 +11,10 @@ readme = "readme.md"
 categories = ["os::windows-apis"]
 exclude = ["tests"]
 
+[features]
+default = []
```
Comment (dpaoliello) [inline crates/libs/registry/Cargo.toml:15]:
```suggestion
default = ["std"]
```

File: crates/libs/registry/Cargo.toml [hunk 3]
```diff
@@ -11,6 +11,10 @@ readme = "readme.md"
...
+std = []
```
Comment (dpaoliello) [inline crates/libs/registry/Cargo.toml:16]:
```suggestion
std = ["windows-result/std"]
```

**Review by dpaoliello — CHANGES_REQUESTED** [review]

File: crates/libs/registry/Cargo.toml (hunk 3 above)
Comment (sivadeilra) [reply crates/libs/registry/Cargo.toml:16]: Actually, the more I think about feature injection, I think we don't actually need it unless A-->B where A actually uses something from B's std-dependent exterior API.

File: crates/libs/registry/Cargo.toml (hunk 2 above)
Comment (sivadeilra) [reply crates/libs/registry/Cargo.toml]: thanks; fixed

File: crates/libs/registry/Cargo.toml (hunk 3 above)
Comment (dpaoliello) [reply crates/libs/registry/Cargo.toml:16]: In which case, we should `check` each crate with no default features to make sure we don't miss dependencies

File: crates/libs/registry/src/value_iterator.rs [hunk 4]
```diff
@@ -79,7 +80,8 @@ impl<'a> Iterator for ValueIterator<'a> {
                             Value::String(String::new())
                         } else {
                             let value = unsafe {
-                                std::slice::from_raw_parts(
+                                // TODO: Vec<u8> does not guarantee alignment for u16
```
Comment (kennykerr) [inline crates/libs/registry/src/value_iterator.rs:83]: Let's create an issue to avoid forgetting about this TODO.

Comment (kennykerr) [reply crates/libs/registry/src/value_iterator.rs:83]: #3048

File: crates/libs/registry/Cargo.toml (hunk 3 above)
Comment (kennykerr) [reply crates/libs/registry/Cargo.toml:16]: * I think `windows-registry` could be entirely `no_std`.
* I could separately add a generated yml workflow for testing no default feature builds.

File: crates/libs/registry/src/value_iterator.rs (hunk 4 above)
Comment (sivadeilra) [reply crates/libs/registry/src/value_iterator.rs:83]: Thanks!

File: crates/libs/registry/Cargo.toml (hunk 3 above)
Comment (sivadeilra) [reply crates/libs/registry/Cargo.toml:16]: Yeah, I see what you mean.  I'm a bit concerned at the combinatorial explosion in `A{no_std?} --> B{no_std?}` variations, so I'm not sure how to implement this suggestion without things blowing up.  But it's definitely a valid concern...

Comment (kennykerr) [reply crates/libs/registry/Cargo.toml:16]: So I think the `windows-registry` crate doesn't need features of its own, depends on the `windows-result` crate with `default-features = false`, and can be `#![no_std]`.

Comment (sivadeilra) [reply crates/libs/registry/Cargo.toml]: Agreed.  I've removed the feature from `windows-registry`; it is now always `no_std` (except in `test` builds).

**Review by kennykerr — APPROVED** [review]

File: crates/libs/registry/Cargo.toml (hunk 3 above)
Comment (kennykerr) [reply crates/libs/registry/Cargo.toml]: Looks like this has been addressed with some follow-up work. Will merge now to get this batch of changes in.

**Pull request merged**